lib/jsonb_accessor/macro.rb

```diff
@@ -61,23 +66,25 @@ def jsonb_accessor(jsonb_attribute, field_types)
           names_and_store_keys.each do |name, store_key|
             define_method("#{name}=") do |value|
               super(value)
-              new_values = (public_send(jsonb_attribute) || {}).merge(store_key => public_send(name))
+              new_values = is_array ? public_send(jsonb_attribute) || names_and_defaults[store_key] : (public_send(jsonb_attribute) || {}).merge(store_key => public_send(name))
               write_attribute(jsonb_attribute, new_values)
             end
           end
 
           # Overrides the jsonb attribute setter to make sure the jsonb fields are kept in sync.
-          define_method("#{jsonb_attribute}=") do |given_value|
-            value = given_value || {}
-            names_to_store_keys = self.class.public_send(store_key_mapping_method_name)
+          unless is_array
```

Comment:
When it is a top level array the name and jsonb_attribute are the same so `name=` method defined above will be overridden by `jsonb_attribute=` method here so we dont execute this block.